CHANGELOG: Move changelogs into a subdir to delegate releng approvals

justaugustus opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Today, changelog updates require top-level approval.
As the Release Engineering subproject and Release Notes roles primarily update the changelog, they should have approval there.

This moves the changelogs into their own directory with an OWNERS file to delegate approval.
Also, we rename the primary `CHANGELOG.md` to `CHANGELOG/README.md` and soft-link it back to its' original location to prevent link breakage.

/assign @liggitt @smarterclayton 
/cc @tpepper @saschagrunert @hasheddan @evillgenius75 @JamesLaverack @puerco 
cc: @kubernetes/release-engineering @kubernetes/release-team 
/sig release
/area release-eng
/priority important-soon
/kind cleanup

**Special notes for your reviewer**:
Something similar was suggested by @liggitt in a previous issue or PR, but I can't locate it at the moment.
/hold to touch base with those working on release notes, as this will change commit locations in the tooling

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

justaugustus committed:
CHANGELOG: Move changelogs into a subdir to delegate releng approvals

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

justaugustus committed:
CHANGELOG: Move changelog, soft-link to top-level, refresh listing

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

liggitt commented:
I'm strongly in favor of this. Did you actually move the existing files? This diff doesn't show them removed. How does github render the softlink?

justaugustus commented:
@liggitt -- I did a `git mv`, but GH sometimes has issues interpreting moves on files with large diffs and across multiple commits.

As for the soft links, they present as a file with a single line naming the link target:
![Screenshot from 2020-02-06 00-54-34](https://user-images.githubusercontent.com/567897/73909739-6d4a8900-487b-11ea-9fe2-1399b8ff9376.png)

For the typechange, locally it will look like this:
```shell
$ git status
On branch changelog
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	typechange: CHANGELOG.md
	new file:   CHANGELOG/README.md
```

...and had it been a typechange with no content change, it might've looked like this on GitHub:
```shell
mode change 100644 => 120000
```

justaugustus commented:
And a diff:
```shell
$ git diff HEAD~2 --name-status 
T       CHANGELOG.md
A       CHANGELOG/CHANGELOG-1.10.md
A       CHANGELOG/CHANGELOG-1.11.md
A       CHANGELOG/CHANGELOG-1.12.md
A       CHANGELOG/CHANGELOG-1.13.md
A       CHANGELOG/CHANGELOG-1.14.md
A       CHANGELOG/CHANGELOG-1.15.md
A       CHANGELOG/CHANGELOG-1.16.md
A       CHANGELOG/CHANGELOG-1.17.md
A       CHANGELOG/CHANGELOG-1.18.md
A       CHANGELOG/CHANGELOG-1.2.md
A       CHANGELOG/CHANGELOG-1.3.md
A       CHANGELOG/CHANGELOG-1.4.md
A       CHANGELOG/CHANGELOG-1.5.md
A       CHANGELOG/CHANGELOG-1.6.md
A       CHANGELOG/CHANGELOG-1.7.md
A       CHANGELOG/CHANGELOG-1.8.md
A       CHANGELOG/CHANGELOG-1.9.md
A       CHANGELOG/OWNERS
A       CHANGELOG/README.md
```

liggitt commented:
> @liggitt -- I did a `git mv`, but GH sometimes has issues interpreting moves on files with large diffs and across multiple commits.

when I check out this branch, I still see the changelogs in the root of the repo, and in the CHANGELOG folder... are you sure you moved them?

justaugustus commented:
> when I check out this branch, I still see the changelogs in the root of the repo, and in the CHANGELOG folder... are you sure you moved them?

Heh. Thanks for checking my work. Looks better now:
```shell
$ git diff HEAD~2 --name-status 
T       CHANGELOG.md
R100    CHANGELOG-1.10.md       CHANGELOG/CHANGELOG-1.10.md
R100    CHANGELOG-1.11.md       CHANGELOG/CHANGELOG-1.11.md
R100    CHANGELOG-1.12.md       CHANGELOG/CHANGELOG-1.12.md
R100    CHANGELOG-1.13.md       CHANGELOG/CHANGELOG-1.13.md
R100    CHANGELOG-1.14.md       CHANGELOG/CHANGELOG-1.14.md
R100    CHANGELOG-1.15.md       CHANGELOG/CHANGELOG-1.15.md
R100    CHANGELOG-1.16.md       CHANGELOG/CHANGELOG-1.16.md
R100    CHANGELOG-1.17.md       CHANGELOG/CHANGELOG-1.17.md
R100    CHANGELOG-1.18.md       CHANGELOG/CHANGELOG-1.18.md
R100    CHANGELOG-1.2.md        CHANGELOG/CHANGELOG-1.2.md
R100    CHANGELOG-1.3.md        CHANGELOG/CHANGELOG-1.3.md
R100    CHANGELOG-1.4.md        CHANGELOG/CHANGELOG-1.4.md
R100    CHANGELOG-1.5.md        CHANGELOG/CHANGELOG-1.5.md
R100    CHANGELOG-1.6.md        CHANGELOG/CHANGELOG-1.6.md
R100    CHANGELOG-1.7.md        CHANGELOG/CHANGELOG-1.7.md
R100    CHANGELOG-1.8.md        CHANGELOG/CHANGELOG-1.8.md
R100    CHANGELOG-1.9.md        CHANGELOG/CHANGELOG-1.9.md
A       CHANGELOG/OWNERS
A       CHANGELOG/README.md
```

liggitt commented on CHANGELOG/OWNERS:
```diff
@@ -0,0 +1,11 @@
+  - sig-release-approvers
+  - release-engineering-approvers
+reviewers:
+  - release-engineering-reviewers
```
probably also docs reviewers/approvers?

justaugustus replied:
Makes sense, but I don't think it's strictly necessary. 
I'll leave it to SIG Docs to make the call on whether they'd like to be added.
cc: @kubernetes/sig-docs-en-owners

zacharysarah replied:
What's your timeline on approval? I'd love to discuss this with SIG Docs at next Tuesday's weekly meeting, if that's OK.

justaugustus commented:
/retest
/assign @zacharysarah @Bradamant3 @jimangel

justaugustus commented:
/retest

saschagrunert reviewed (approved):
/lgtm

We have to adapt the release engineering tooling to write to the new paths.

justaugustus commented:
/retest

justaugustus commented:
/retest

saschagrunert commented on CHANGELOG/README.md:
```diff
@@ -0,0 +1,29 @@
+
+- [CHANGELOG-1.17.md](./CHANGELOG-1.17.md)
+
+## Older releases
```
Do we really need that file? My concern is that it is hard to automate via `krel` to update that file since we have those 3 sections. We could also stick to a plain list here…

justaugustus replied:
@saschagrunert -- I'd like to keep the file as it's likely referenced in a few places outside of our repo. Agreed that it'll be easier to manage if we do a single list though and I've made that change in e54bfa4652d2f87c7c625bd2758a96a1fb8c4c7c.

saschagrunert replied:
Sounds good, thanks :pray:

justaugustus committed:
CHANGELOG: Collapse README.md headings in single list

While it's not happening today, we'll want to eventually also handle
updates to this CHANGELOG list via krel changelog. By collapsing the set
of changelogs into a single list, it will be a little easier to write
out an updated file for commit by searching the contents of the directory
that match "CHANGELOG-x.y.md" and adding them to the list.

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

justaugustus commented:
/retest

justaugustus commented:
/retest

justaugustus commented:
/retest

hasheddan reviewed (approved):
/lgtm

justaugustus commented:
Update in-flight on the k/release side (https://github.com/kubernetes/release/pull/1064), so I'm lifting the hold.
/hold cancel

xmudrii reviewed (approved):
/lgtm

puerco commented:
/lgtm
I don't think this change affects the paths used in the current release notes process as only the changelog files are shifted into the new directory

liggitt commented:
> Update in-flight on the k/release side ([kubernetes/release#1064](https://github.com/kubernetes/release/pull/1064)), so I'm lifting the hold.

is the same tool used against master and release branches? do we need to make an identical change in release branches or does the tool tolerate both locations?

saschagrunert commented:
> > Update in-flight on the k/release side ([kubernetes/release#1064](https://github.com/kubernetes/release/pull/1064)), so I'm lifting the hold.
> 
> is the same tool used against master and release branches? 

Yes, but patch releases are handled in a different way right now. I will fix that as well in the mentioned PR.

> do we need to make an identical change in release branches or does the tool tolerate both locations?

Good point, I think it would be good if we can change that for supported patch-release branches as well.

tpepper commented:
I like the idea, ie:
https://kubernetes.slack.com/archives/CJH2GBF7Y/p1572545069093000

Back then there was discussion of whether we could just move the files or if we needed to tombstone them.  Since the changelogs are not designed for machine readability, I'd be surprised if machines were often reading them.  Probably something somewhere is slightly broken.  But I would like to think a human would quickly and easily observe the like named subdirectory and things would get resolved quickly and easily by using that subdir instead of the root directory.  But this disruption is something to consider, I think especially were this patch merged and then proposed as a cherry pick on today's patch release branches.

zacharysarah commented:
This makes sense. 

I'm a strong +1, with a request to delay until 12pm on 11 Feb so SIG Docs can communally review whether we should have a reviewer entry in the OWNERS file.

liggitt commented:
/approve

justaugustus commented:
/retest

JamesLaverack reviewed (approved):
/lgtm
